### PR TITLE
Fix perf crashes in RPS and HPS tests

### PR DESCRIPTION
**Pull request by ThadHouse**
Each had a different style of race that was either resulting in a use after free (HPS) or a double free (RPS)

Fixing the double free requires an API change, so that in a shutdown complete event, it can be determined if it was app
initiated or connection initiated

### CODE REVIEW

File: src/core/connection.h [hunk 1]
```diff
@@ -147,6 +147,12 @@ typedef union QUIC_CONNECTION_STATE {
         //
         BOOLEAN ResumptionEnabled : 1;
 
+        //
+        // Indicates that an app shutdown from a non worker thread is in progress.
+        // Received by the QUIC_CONNECTION_EVENT_SHUTDOWN_COMPLETE event.
+        //
+        BOOLEAN AppShutdownInProgress: 1;
```
Comment (nibanks) [inline src/core/connection.h:154]: "Shutdown" is a different concept. It's in Close right now. Call it `AppCloseInProgress` instead.

File: src/inc/msquic.h [hunk 2]
```diff
@@ -786,6 +786,7 @@ typedef struct QUIC_CONNECTION_EVENT {
         struct {
             BOOLEAN HandshakeCompleted;
             BOOLEAN PeerAcknowledgedShutdown;
+            BOOLEAN AppShutdownInProgress;
```
Comment (nibanks) [inline src/inc/msquic.h:789]: Let's go ahead and make this a bit field now that we have several.

File: src/perf/lib/HpsClient.h [hunk 3]
```diff
@@ -33,7 +33,7 @@ struct HpsWorkerContext {
         QuicZeroMemory(&LocalAddrs, sizeof(LocalAddrs));
         QuicEventInitialize(&WakeEvent, FALSE, TRUE);
     }
-    ~HpsWorkerContext() {
+    void WaitForWorker() {
         if (ThreadStarted) {
```
Comment (nibanks) [inline src/perf/lib/HpsClient.h:37]: Don't do it this way. Instead make a new WaitForWorker function that just sets the event and waits on the thread. Keep the destructor as is.

File: src/inc/msquic.h [hunk 4]
```diff
@@ -784,8 +784,9 @@ typedef struct QUIC_CONNECTION_EVENT {
             QUIC_UINT62 ErrorCode;
         } SHUTDOWN_INITIATED_BY_PEER;
         struct {
-            BOOLEAN HandshakeCompleted;
-            BOOLEAN PeerAcknowledgedShutdown;
+            BOOLEAN HandshakeCompleted          : 1;
+            BOOLEAN PeerAcknowledgedShutdown    : 1;
+            BOOLEAN AppCloseInProgress       : 1;
```
Comment (nibanks) [inline src/inc/msquic.h:789]: nit: spacing

**Review by nibanks — APPROVED** [review]

**Review by nibanks — APPROVED** [review]

**Comment by ThadHouse** [comment]
Closing just so I can run perf pipeline. Will reopen.

**Review by nibanks — APPROVED** [review]

**Pull request merged**